### DISCUSSION TITLE
Why there is no mdx highlight in treesitter?

### DISCUSSION THREAD

**Discussion by theaungmyatmoe** [nvim-treesitter/nvim-treesitter#3908] [Q&A]
I want to highlight MDX in nvim. But ts didn't support syntax highlight or parsing for mdx.

**Comment by clason** [comment]
Nobody contributed a parser and queries, that's why. (Maybe _you_ can do it?)

**Reply by theaungmyatmoe** [reply]
I may be contribute in future, buy currently busy with full stack development

**Comment by mrded** [comment]
Better than nothing:
```
vim.filetype.add({
  extension = { mdx = 'mdx' }
})

vim.treesitter.language.register('mdx', 'markdown')
```

**Comment by DefectingCat** [comment]
Also you can create a autocmd for set `*.mdx` file as markdown file to get a little bit hight

```lua
-- set markdown highlight for mdx file
autocmd({ "BufNewFile", "BufRead" }, {
  pattern = { "*.mdx" },
  callback = function()
    local buf = vim.api.nvim_get_current_buf()
    vim.api.nvim_buf_set_option(buf, "filetype", "markdown")
  end,
})
```

**Comment by griff-rees** [comment]
An example by [`phelipetls`](https://github.com/phelipetls): https://phelipetls.github.io/posts/mdx-syntax-highlight-treesitter-nvim/

**Comment by LittleLegend-Pengiun** [comment]
# NvChad Guide

This is the guide for someone who use NvChad, like me. 

## Step 1

Make sure you are in the `nvim` config directory, the place where you install all your NvChad config. We only work in this directory from now on.

## Step 2

Open the `nvim/init.lua` file, navigate to the top, and copy the lines below. It MUST BE ON TOP, or before importing all the other nvim plugin, otherwise it will think everything on screen is markdown.

```
vim.api.nvim_create_autocmd({ "BufNewFile", "BufRead" }, {
	pattern = { "*.mdx" },
	callback = function()
		local buf = vim.api.nvim_get_current_buf()
		vim.api.nvim_buf_set_option(buf, "filetype", "markdown")
	end,
})
```
Or if you prefer something simpler, but that won't activate other markdown filetype plugins you may want.

```
vim.api.nvim_create_autocmd({ "BufNewFile", "BufRead" }, {
	pattern = { "*.mdx" },
	callback = function()
		vim.treesitter.start(buf, 'markdown')
	end,
})
```

## Step 3

**Notes**: Things still work if you put the queries in `/nvim/queries/markdown/**`.

In the `nvim` directory, create the file `nvim/after/queries/markdown/injections.scm` and put the following content inside:
```
; extends
((inline) @injection.content
  (#lua-match? @injection.content "^%s*import")
  (#set! injection.language "typescript"))
((inline) @injection.content
  (#lua-match? @injection.content "^%s*export")
  (#set! injection.language "typescript"))
```

Repeat the process with another file name `nvim/after/queries/markdown/highlights.scm`:
```
; extends
((inline) @_inline (#lua-match? @_inline "^%s*import")) @nospell
((inline) @_inline (#lua-match? @_inline "^%s*export")) @nospell
```

And there you have it, next time you open nvim with a mdx file, everything is highlighted as you expected.


![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/65987402/76152d6d-3018-4be4-91ea-4275b7db8889)

**Reply by clason** [reply]
I don't see anything NvChad specific here, though. And the queries should go on `/nvim/queries`, not `after/`. Otherwise looks good!

**Reply by LittleLegend-Pengiun** [reply]
The NvChad has tone of stub pre-configured in the `init.lua`. So if you put the config after the neovim loading all the plugins, it won't work. Some new users, like I am, don't know where to put these stubs in properly. About the queries, I'll try your guide to see whether it works or not, and then update the guide if needed. Thank you.

**Reply by clason** [reply]
Also, your autocommand in Step 2 conflicts with the two commands above it. Either you `add` MDX as a new filetype and register the markdown parser for it, or you detect `.mdx` files directly as Markdown (which you can also do with `vim.filetype.add`, which creates these autocommands under the hood.

You can also simply add an autocommand that calls `vim.treesitter.start(buf, 'markdown')` for `*.mdx` files (although that won't activate other markdown filetype plugins you may want).

It all depends on whether MDX files are "proper" Markdown files with extras and should be treated as such, or whether you _just_ want to run the tree-sitter parser on MDX files.

**Reply by LittleLegend-Pengiun** [reply]
Thanks for your comment, I think I'll stick with the second option.

**Reply by linsjb** [reply]
This is awesome. Just what I needed. Thank you! 🙏
